### ISSUE TITLE
Severe performance regression in 3.8.14

### ISSUE THREAD

**Issue by Sebmaster** [Automattic/mongoose#4821]
The fix for #2140 (cdd63bf6ff0854018cd6b70fd8397c9e942dcd59) introduced a quite severe performance regression.

Compare the runtime of this code sample with mongoose 3.8.13 and 4.7.2:
```js
var mongoose = require('mongoose')
var fs = require('fs')

var Schema = mongoose.Schema;

ChecklistSchema = new Schema({
    name: { type: String }
});

BoardSchema = new Schema({
    checklists: { type: [ChecklistSchema] }
});

var Board = mongoose.model('Board', BoardSchema);

var doc = new Board();
for (var i = 0; i < 1000; ++i) {
    var checklist = {name:'test'};
    doc.checklists.push(checklist);
}

var obj = doc.toObject();
var time2 = 0;
var start2 = new Date();
for (var i = 0; i < 100; i++) {
  new Board(obj);
}
console.error('instantiating schema object took %d ms', (new Date - start2));
```

```
sebastian@host:~/Desktop/test$ node big # mongoose 3.8.13
instantiating schema object took 1127 ms
sebastian@host:~/Desktop/test$ node big # mongoose 4.7.3
instantiating schema object took 6898 ms
sebastian@host:~/Desktop/test$ node big # mongoose 4.7.3 with commented try-catch
instantiating schema object took 228 ms
```

So at this point I think it'd be better to either make `.id` properly handle a null subdoc, or alternatively loop through all subdocs before calling id and checking for null first (and then avoid calling `.id` if any are found).

Node v6.9.2
mongoose 4.7.3

**Comment by Sebmaster** [comment]
Digging further into this, it actually seems like we might be hitting a v8 deopt due to a too long method body. Replacing the try-catch with a simple `if (prev)` keeps the slowdown. Removing the if makes it fast.

**Comment by Sebmaster** [comment]
Okay, after digging some more: just dropping the try-catch just makes everything throw, so does a lot less work. Replacing the try-catch with an explicit filter for null does lead to some improvement though, still does not make a dent really though.

Most of the time spent is actually in `Document.prototype.__registerHooksFromSchema`, dropping the call to that improves drops the time to ~1400ms which is around what mongoose 3.8 took. But in the current state, creating lots of documents with mongoose is way too slow :disappointed:

**Comment by brettkiefer** [comment]
It looks like more of a gradual thing than a single problem:

`for release in 3.2.2 3.8.40 4.0.3 4.7.3; do git checkout $release && npm install > /dev/null && node big.js ; done`

3.2.2 - instantiating schema object took 2717 ms
3.8.40 - instantiating schema object took 6693 ms
4.0.3 - instantiating schema object took 10496 ms
4.7.3 - instantiating schema object took 14944 ms

It looks like the performance problem must be in $__registerHooksFromSchema, so maybe related to https://github.com/Automattic/mongoose/issues/2754 ? Maybe largely related to changes to $__registerHooks to support promises?

**Comment by brettkiefer** [comment]
From this test it looks like the expensive part is https://github.com/Automattic/mongoose/blob/master/lib/document.js#L2017 and https://github.com/Automattic/mongoose/blob/master/lib/document.js#L2021, so that seems to point to hooks-fixed?

**Comment by Sebmaster** [comment]
@vkarpov15 Is the plan still to move over to Kareem, or is it grappling-hook now? Would it make sense to dig into hooks-fixed to see if there's some temporary workaround to improve performance for now?

**Comment by vkarpov15** [comment]
Not decided between kareem, grappling-hook, or something else, suggestions welcome on that front.

Yeah, definitely would make sense to dig into hooks-fixed. I'll look into this over the next couple days and see if there's any solution I can come up with.

**Comment by vkarpov15** [comment]
So above commit removes the offending try/catch and shaves ~25% off of your example:

```
$ node gh-4821.js 
instantiating schema object took 8542 ms
$ node gh-4821.js 
instantiating schema object took 6587 ms
```

**Comment by brettkiefer** [comment]
Neat! That's great, it seems to get back down to about the same performance as 4.0.3; I guess that performance difference between 3.2.2 and 3.8.40 is a lot of what that change gets rid of, since the try/catch was introduced in 3.8.14? It seems like a lot of the rest of the performance hit comes from the changes to set up promise support for the callbacks; I suppose it would be a weird optimization to allow the client to opt out of that and do its own hook promisification.

**Comment by vkarpov15** [comment]
So I shaved another 500ms off:

```
$ node gh-4821.js 
instantiating schema object took 6004 ms
$ 
```

The problem now is mostly the fact that $__registerHooksFromSchema executes every single time a document is instantiated, rather than just modifying the model used to construct the document once, which came from the promise changes that you mentioned.

I'll see if I can refactor it to make it less painfully slow for 4.8, because as an unfortunate side effect of $__registerHooksFromSchema running every time a document is created, adding save hooks to a schema after creating a model works. No matter how much we say this behavior is not supported in the docs, somebody inevitably is using it, so better to at least put it in a patch release.

**Comment by brettkiefer** [comment]
That checks out for me, I also get about another 500ms back (with 7006b5ec86b19a5b1df8618b2116c8810f173d0f, not f3e9400a7aa9f8e12b28c20fa49e41658c600e40, but the one seems to be an extension of the other?)

**Comment by vkarpov15** [comment]
Yeah they're pretty similar. General principle is that forEach is very slow so removing that gives us a lot back. Making it so that registerHooksFromSchema doesn't run on every document will make it much faster in theory

**Comment by vkarpov15** [comment]
Using the head of [this branch](https://github.com/Automattic/mongoose/tree/gh-2754-2) as it stands right now is now much faster than 3.8.40, will release that as part of 4.8

```
$ node gh-4821.js 
instantiating schema object took 2337 ms
$
$ rm -rf node_modules/mongoose && npm install mongoose@3.8.40
$ node gh-4821.js 
js-bson: Failed to load c++ bson extension, using pure JS version
instantiating schema object took 4490 ms
$ 
```

:rocket:

**Comment by brettkiefer** [comment]
That all checks out great in my tests!

**Comment by Sebmaster** [comment]
@vkarpov15 That branch looks awesome in our prod tests, any ETA on the release of 4.8?

**Comment by vkarpov15** [comment]
Hopefully end of this week or start of next week, have [a few more issues to wrap up](https://github.com/Automattic/mongoose/milestone/88)
